metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java

```diff
@@ -0,0 +1,894 @@
+                    responseTopicData.partitions().add(new AlterIsrResponseData.PartitionData().
+                        setPartitionIndex(partitionData.partitionIndex()).
+                        setErrorCode(Errors.INVALID_REQUEST.code()));
+                }
+                records.add(new ApiMessageAndVersion(new PartitionChangeRecord().
+                    setPartitionId(partitionData.partitionIndex()).
+                    setTopicId(topic.id).
+                    setIsr(partitionData.newIsr()), (short) 0));
+            }
+        }
+        return new ControllerResult<>(records, response);
+    }
+
+    /**
+     * Generate the appropriate records to handle a broker being fenced.
+     *
+     * First, we remove this broker from any non-singleton ISR. Then we generate a
+     * FenceBrokerRecord.
+     *
+     * @param brokerId      The broker id.
+     * @param records       The record list to append to.
+     */
+
+    void handleBrokerFenced(int brokerId, List<ApiMessageAndVersion> records) {
+        BrokerRegistration brokerRegistration = clusterControl.brokerRegistrations().get(brokerId);
+        if (brokerRegistration == null) {
+            throw new RuntimeException("Can't find broker registration for broker " + brokerId);
+        }
+        handleNodeDeactivated(brokerId, false, records);
+        records.add(new ApiMessageAndVersion(new FenceBrokerRecord().
+            setId(brokerId).setEpoch(brokerRegistration.epoch()), (short) 0));
+    }
+
+    /**
+     * Generate the appropriate records to handle a broker being unregistered.
+     *
+     * First, we remove this broker from any non-singleton ISR. Then we generate an
+     * UnregisterBrokerRecord.
+     *
+     * @param brokerId      The broker id.
+     * @param brokerEpoch   The broker epoch.
+     * @param records       The record list to append to.
+     */
+    void handleBrokerUnregistered(int brokerId, long brokerEpoch,
+                                  List<ApiMessageAndVersion> records) {
+        handleNodeDeactivated(brokerId, false, records);
+        records.add(new ApiMessageAndVersion(new UnregisterBrokerRecord().
+            setBrokerId(brokerId).setBrokerEpoch(brokerEpoch), (short) 0));
+    }
+
+    /**
+     * Handle a broker being deactivated. This means we remove it from any ISR that has
+     * more than one element. We do not remove the broker from ISRs where it is the only
+     * member since this would preclude clean leader election in the future.
+     * It is removed as the leader for all partitions it leads.
+     *
+     * @param brokerId              The broker id.
+     * @param alwaysBumpLeaderEpoch True if we should generate a record that
+     *                              always increments the leader epoch.
+     * @param records               The record list to append to.
+     */
+    void handleNodeDeactivated(int brokerId, boolean alwaysBumpLeaderEpoch,
```

Comment:
Currently, for controller initiated ISR change (controlled shutdown or hard failure), we always bump up the leader epoch. Also, the name alwaysBumpLeaderEpoch is a bit weird since the code in handleNodeDeactivated() doesn't directly bump up leader epoch.

Reply:
Hmm... I don't completely understand why we would want to bump the leader epoch when the controller removes a non-leader broker B but not when an AlterIsrRequest removes a non-leader broker B from the ISR.  It seems like we should either bump in both scenarios or neither.

Is the fact that we bump in the first scenario just an artifact of the fact that otherwise we could not send out a leader and isr request that had a new epoch and thereby caused a  change?

Reply:
As per our discussion outside github, let's just use the old behavior for now.